KMB/make_KMB_info.py

```diff
@@ -0,0 +1,270 @@
+
+Transforms the partially processed data from kmb_massload into a
+BatchUploadTools compliant json file.
+"""
+from __future__ import unicode_literals
+from collections import OrderedDict
+
+import pywikibot
+import batchupload.helpers as helpers
+import batchupload.common as common
+from batchupload.make_info import MakeBaseInfo
+
+
+BATCH_CAT = 'Media contributed by RAÄ'  # stem for maintenance categories
+BATCH_DATE = '2017-05'  # branch for this particular batch upload
+
+
+class KMBInfo(MakeBaseInfo):
+    """Construct file descriptions and filenames for the KMB batch upload."""
+
+    def __init__(self, **options):
+        """Initialise a make_info object."""
+        super(KMBInfo, self).__init__(BATCH_CAT, BATCH_DATE, **options)
+
+    def load_data(self, in_file):
+        """
+        Load the provided data (output from kmb_massload).
+
+        Return this as a dict with an entry per file which can be used for
+        further processing.
+
+        :param in_file: the path to the metadata file
+        :return: dict
+        """
+        return common.open_and_read_file(in_file, as_json=True)
+
+    # @todo: Not all problems should necessarily result in skipping the image
+    #        completely. And some other issues possibly should - T164578
+    def process_data(self, raw_data):
+        """
+        Take the loaded data and construct a KMBItem for each.
+
+        Populates self.data
+
+        :param raw_data: output from load_data()
+        """
+        d = {}
+        for key, value in raw_data.iteritems():
+            item = KMBItem(value)
+            if item.problem:
+                pywikibot.output(
+                    'The {0} image was skipped because of: {1}'.format(
+                        item.ID, item.problem))
+            else:
+                d[key] = item
+
+        self.data = d
+
+    # @todo: break out substed lists as mappings - T164567
+    def load_mappings(self, update=True):
+        """
+        Update mapping files, load these and package appropriately.
+
+        :param update: whether to first download the latest mappings
+        """
+        # Currently these are handled as substed templates.
+        # redo as either:
+        # * offline mappings: If so consider re-jiging make_info to not assume
+        #   it is alway an online table.
+        # * online tables: If so consider re-jiging make_info to not assume
+        #   these need alway be dumped a offline files first.
+        pass
+
+    # @note: this differs from the one created in RAA-tools
+    def generate_filename(self, item):
+        """
+        Given an item (dict) generate an appropriate filename.
+
+        The filename has the shape: descr - Collection - id
+        and does not include filetype
+
+        :param item: the metadata for the media file in question
+        :return: str
+        """
+        return helpers.format_filename(item.get_description(), 'KMB', item.ID)
```

Comment:
This is checked during the actual upload. The code for this is in BatchUploadTools. In short traditionally all of the images+descriptions are in one directory. They then get filtered down into uploaded, warning, error subdirectories depending on if the upload worked.

If the filename already exists or (more likely) the image was already uploaded under a different name this will trigger a warning/error, be logged and the image moved to the appropriate subdirectory.

Now the whole all images in the same directory assumption has to be addressed as part of [T158402](https://phabricator.wikimedia.org/T158402)